go/feast/featurestore.go

```diff
@@ -0,0 +1,907 @@
+	requestFvsToUse := make([]*RequestFeatureView, len(requestFvs))
+	odFvsToUse := make([]*OnDemandFeatureView, len(odFvs))
+	index := 0
+	for _, fv := range fvs {
+		fvsToUse[index] = fv
+		index += 1
+	}
+
+	index = 0
+	for _, fv := range requestFvs {
+		requestFvsToUse[index] = fv
+		index += 1
+	}
+	index = 0
+	for _, fv := range odFvs {
+		odFvsToUse[index] = fv
+		index += 1
+	}
+	return fvsToUse, requestFvsToUse, odFvsToUse, nil
+}
+
+func (fs *FeatureStore) getEntityMaps(requestedFeatureViews []*FeatureView) (map[string]string, error) {
+
+	entityNameToJoinKeyMap := make(map[string]string)
+	var entityNames map[string]bool
+	var entityName string
+	var joinKeyMap map[string]string
+	var featureView *FeatureView
+
+	entities, err := fs.listEntities(true, false)
+
+	if err != nil {
+		entities = make([]*Entity, 0)
+	}
+
+	for _, entity := range entities {
+		entityNameToJoinKeyMap[entity.name] = entity.joinKey
+	}
+
+	for _, featureView = range requestedFeatureViews {
+
+		entityNames = featureView.entities
+		joinKeyMap = featureView.base.projection.joinKeyMap
+		for entityName = range entityNames {
+
+			entity, err := fs.registry.getEntity(fs.config.Project, entityName)
+			if err != nil {
+				return nil, err
+			}
+			entityName := entity.name
+			joinKey := entity.joinKey
+
+			// TODO (Ly): Review: weird that both uses the same map?
+			// from python's sdk
+			if entityNameMapped, ok := joinKeyMap[joinKey]; ok {
+				entityName = entityNameMapped
+			}
+			if joinKeyMapped, ok := joinKeyMap[joinKey]; ok {
+				joinKey = joinKeyMapped
+			}
+			entityNameToJoinKeyMap[entityName] = joinKey
+			// TODO (Ly): Review: Can we skip entity_type_map
+			// in go's version?
+		}
+	}
+	return entityNameToJoinKeyMap, nil
+}
+
+func (fs *FeatureStore) validateEntityValues(joinKeyValues map[string]*types.RepeatedValue) (int, error) {
+	setOfRowLengths := make(map[int]bool)
+	var numRows int
+	for _, col := range joinKeyValues {
+		setOfRowLengths[len(col.Val)] = true
+		numRows = len(col.Val)
+	}
+	if len(setOfRowLengths) > 1 {
+		return 0, errors.New("ValueError: All entity rows must have the same columns.")
+	}
+	return numRows, nil
+}
+
+func (fs *FeatureStore) validateFeatureRefs(featureRefs []string, fullFeatureNames bool) error {
+	collidedFeatureRefs := make(map[string]int)
+	if fullFeatureNames {
+		for _, featureRef := range featureRefs {
+			collidedFeatureRefs[featureRef] += 1
+		}
+		for featureName, occurrences := range collidedFeatureRefs {
+			if occurrences == 1 {
+				delete(collidedFeatureRefs, featureName)
+			}
+		}
+		if len(collidedFeatureRefs) >= 1 {
+			collidedFeatureRefList := make([]string, len(collidedFeatureRefs))
+			index := 0
+			for featureName := range collidedFeatureRefs {
+				collidedFeatureRefList[index] = featureName
+				index += 1
+			}
+			return errors.New(fmt.Sprintf("FeatureNameCollisionError: %s; %t", strings.Join(collidedFeatureRefList, ", "), fullFeatureNames))
```

Comment:
Instead of using strings for error names, would it be better to use custom errors? I also think that just using an error name here is a bit cryptic. I would probably just start with an error message.